src/cpp/core/file_lock/FileLock.cpp

```diff
@@ -0,0 +1,257 @@
+{
+   double value = settings.getDouble(name, defaultValue);
+   if (value < 0)
+   {
+      LOG_WARNING_MESSAGE("invalid field '" + name + "': must be positive");
+      return defaultValue;
+   }
+   
+   return value;
+}
+
+bool s_isInitialized = false;
+
+} // end anonymous namespace
+
+bool FileLock::verifyInitialized()
+{
+   if (!s_isInitialized)
+   {
+      static bool s_warned = false;
+      if (s_warned)
+         return s_isInitialized;
+      
+      s_warned = true;
+      LOG_WARNING_MESSAGE(
+               "FileLock classes not yet initialized; please call "
+               "'FileLock::initialize()' and 'FileLock::cleanUp()' as appropriate");
+   }
+   
+   return s_isInitialized;
+}
+
+void FileLock::initialize(FilePath locksConfPath)
+{
+   s_isInitialized = true;
+   
+   if (locksConfPath.empty())
+      locksConfPath = FilePath(kLocksConfPath);
+   
+   if (!locksConfPath.exists())
+      return;
+   
+   Settings settings;
+   Error error = settings.initialize(locksConfPath);
+   if (error)
+   {
+      LOG_ERROR(error);
+      return;
+   }
+   
+   FileLock::initialize(settings);
+}
+
+void FileLock::initialize(const Settings& settings)
+{
+   s_isInitialized = true;
+   
+   // default lock type
+   FileLock::s_defaultType = stringToLockType(settings.get("lock-type", kLockTypeAdvisory));
+   
+   // timeout interval
+   double timeoutInterval = getFieldPositive(settings, "timeout-interval", kDefaultTimeoutInterval);
+   FileLock::s_timeoutInterval = boost::posix_time::seconds(timeoutInterval);
+   
+   // refresh rate
+   double refreshRate = getFieldPositive(settings, "refresh-rate", kDefaultRefreshRate);
+   FileLock::s_refreshRate = boost::posix_time::seconds(refreshRate);
+   
+   DEBUG_BLOCK("lock initialization")
+   {
+      std::cerr << "Type: "    << lockTypeToString(FileLock::s_defaultType) << std::endl;
+      std::cerr << "Timeout: " << FileLock::s_timeoutInterval.total_seconds() << std::endl;
+      std::cerr << "Refresh: " << FileLock::s_refreshRate.total_seconds() << std::endl;
+   }
+}
+
+// default values for static members
+FileLock::LockType FileLock::s_defaultType(FileLock::LOCKTYPE_LINKBASED);
+boost::posix_time::seconds FileLock::s_timeoutInterval(kDefaultTimeoutInterval);
+boost::posix_time::seconds FileLock::s_refreshRate(kDefaultRefreshRate);
+
+boost::shared_ptr<FileLock> FileLock::create(LockType type)
+{
+   verifyInitialized();
+   
+   switch (type)
+   {
+   case LOCKTYPE_ADVISORY:  return boost::shared_ptr<FileLock>(new AdvisoryFileLock());
+   case LOCKTYPE_LINKBASED: return boost::shared_ptr<FileLock>(new LinkBasedFileLock());
+   }
+   
+   // shouldn't be reached
+   return boost::shared_ptr<FileLock>(new AdvisoryFileLock());
+}
+
+boost::shared_ptr<FileLock> FileLock::createDefault()
+{
+   verifyInitialized();
+   return FileLock::create(s_defaultType);
+}
+
+void FileLock::refresh()
+{
+   verifyInitialized();
+   AdvisoryFileLock::refresh();
```

Comment:
The overall goal was for `refresh()` to be part of the `FileLock` 'interface', and all classes implementing `FileLock` should provide `refresh()`.

Of course, since this is a static method, the above is not really true but for consistency I think it's okay to implement / call that in all sub-classes.